### ISSUE TITLE
Move Tycho-Pomless documention from wiki to site

### ISSUE THREAD

**Issue by HannesWell** [eclipse-tycho/tycho#1260]
At the moment a lot of documention about Tycho-Pomless resides at this repos wiki at:
https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless

I think that documentation would be better placed at [Tycho's documentation site](https://www.eclipse.org/tycho/sitedocs/). That's the place where I would search for it first.
If I understand it correctly a corresponding `TychoPomless.md` markdown file can be placed at [src/site/markdown](https://github.com/eclipse-tycho/tycho/tree/master/src/site/markdown) and will eventually show up at Tycho's Site?
If this is correct the content from the wiki could be (more or less?) simply be copied.

This would also have the advantage that the documentation can be maintained together with the code.

**Comment by laeubi** [comment]
Site documentation is only updated on release, I found it more useful to update some docs also between releases.

**Comment by akurtakov** [comment]
You'll have to reference it in https://github.com/eclipse-tycho/tycho/blob/master/src/site/site.xml AFAIU.
@laeubi site seems to not have been published last for 2.7.0. Can you publish the 2.7.4 ?

**Comment by laeubi** [comment]
If we want to use more "site" stuff I think publishing should be automated so they are simply published on every build of a tycho-x-branch.

**Comment by laeubi** [comment]
By the way currently this seem to require a gerrit (!) to be created, we maybe better publish it to the github-pages, especially now we have an won tycho org? See https://pages.github.com/ ... this should allow us to fully automate this.

**Comment by laeubi** [comment]
@LorenzoBettini as you have [described it here](https://www.lorenzobettini.it/2020/01/publishing-a-maven-site-to-github-pages/) do you think you can help us setting this up for tycho if we create the necessary repository?

**Comment by HannesWell** [comment]
> Site documentation is only updated on release, I found it more useful to update some docs also between releases.

That's a point.

> If we want to use more "site" stuff I think publishing should be automated so they are simply published on every build of a tycho-x-branch.

Sounds very good to me. I also wondered if it would be useful to keep the site of previous releases. At least of some. Maybe only for minors or at least the last one of a major branch. I think those that don't want/can migrate to Tycho 3.0 once it is released would be happy to at least have the documentation available.
IIRC I have seen it that there is a drop-down menu next to the version where one can select another version.

Besides the releases (current and some previous?) there could also be a snapshot of the current master so that can can see an update immediately.

**Comment by laeubi** [comment]
I think it should be something like:

- master/
- tycho-2.7.x
- tycho-3.0.x
- ... and so on

We can the use an index file that simply gives a link to the subfolder (unless someone likes to create a page for tycho itself...)

**Comment by HannesWell** [comment]
Sounds good, but I cannot help with the website stuff itself. Never done this.

**Comment by laeubi** [comment]
If @mickaelistria and @akurtakov are fine with that, I'll create a ticket with infra to setup the github pages repo tomorrow... that would be the first step.

**Comment by akurtakov** [comment]
Go :)

**Comment by LorenzoBettini** [comment]
> @LorenzoBettini as you have [described it here](https://www.lorenzobettini.it/2020/01/publishing-a-maven-site-to-github-pages/) do you think you can help us setting this up for tycho if we create the necessary repository?

@laeubi yes, with pleasure! But I'm on holiday, I can get back to you approx next week, ok?

**Comment by laeubi** [comment]
> > @LorenzoBettini as you have [described it here](https://www.lorenzobettini.it/2020/01/publishing-a-maven-site-to-github-pages/) do you think you can help us setting this up for tycho if we create the necessary repository?
> 
> @laeubi yes, with pleasure! But I'm on holiday, I can get back to you approx next week, ok?

I don't think we are in hurry here, so no problem :-)

**Comment by laeubi** [comment]
> Go :)

Here we go https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1650

**Comment by laeubi** [comment]
@LorenzoBettini the page is setup now! So we are ready to accept a PR whenever you are ready :-)

- Web link: https://eclipse-tycho.github.io/
- Repository: https://github.com/eclipse-tycho/eclipse-tycho.github.io

I would suggest we start with the following:

1. There is a `doc` folder in the root
2. for each branch that starts with `tycho-` a sitedoc should be generated and placed inside `doc/<branch name>`
3. we have a GH action that is triggered by a push to any `tycho-*` branch that publishes the sitedoc from this branch into the folder
4. we edit the `index.html` file to have a link to each subfolder (has to be done manually at the moment), if I understand the github docs correctly we can also use a README.MD if that is more sufficient.

The next steps could then to adjust the sitedocs itself, maybe with a link that says "Looking for another Tycho version?" that reference to the index.html page.

**Comment by LorenzoBettini** [comment]
@laeubi, I'm not sure I understand how you want to proceed. My blog post is about publishing a Maven site to GitHub pages, together with plugin documentation. So the actual source repository should be this one.

**Comment by laeubi** [comment]
@LorenzoBettini you can simply push it to https://github.com/eclipse-tycho/eclipse-tycho.github.io master branch instead of the Tycho repository itself.

You can take a look at https://github.com/eclipse-tycho/tycho/blob/master/.github/workflows/verify-platform.yml how to checkout different repositories side-by-side.

**Comment by laeubi** [comment]
@LorenzoBettini if you still think its easier to push it to the same repo, we can still do this and link from the main site to the tycho-project site, so I would laeve the descission to you what is more suitable.

**Comment by mickaelistria** [comment]
IMO, best is to have "manual" documentation directly in the repo, together with the code (same branch, not in a wiki), in some "documentation" folder. Maintaining a website is more complicated and doesn't bring additional value.
As for the automated documentation, how is GitHub a better service than the current eclipse.org site? If there is no additional value to expect, it's probably not worth changing.

**Comment by laeubi** [comment]
> IMO, best is to have "manual" documentation directly in the repo, together with the code (same branch, not in a wiki), in some "documentation" folder. Maintaining a website is more complicated and doesn't bring additional value.

The suggestion was to maintain this as part of the site-doc, so we get a website generated automatically.

> As for the automated documentation, how is GitHub a better service than the current eclipse.org site? If there is no additional value to expect, it's probably not worth changing.

We can deploy automatically the site-docs by GH action, while currently it requires us to generate the site doc locally, commit to a gerrit repo (no idea what will happen to that in the future ...) open a review, and as far as I can remember only some people can then actually merge this.

**Comment by mickaelistria** [comment]
I actually challenge even the fact that we need a website at all. The benefits you're pointing would apply to a documentation folder in the master branch as well.

**Comment by laeubi** [comment]
> I actually challenge even the fact that we need a website at all.

Do we really want to start defining what "a website" is? Is Github UI a "website"? Beside that, Eclipse projects really would benfit from having a "real" website, just take https://karaf.apache.org/ and then look at the "websites" of usual eclipse projects e.g. Tycho https://projects.eclipse.org/projects/technology.tycho I think yes we

1. could really need a website to better explain what tycho is, how to access the docs and examples
2. there are many shades of "website"

> The benefits you're pointing would apply to a documentation folder in the master branch as well.

Only for people who like reading source codes, see above... all content you can browse on the web is actually a website and I assume we don't want people to checkout the source an look at raw text files instead.

A folder in the source code is really hard to discover, hard to link to and has a bad user experience.

**Comment by laeubi** [comment]
> Isn't it actually our target audience?

The target audience are people that need to setup/configure a Tycho build. These might be developers, build managers or whatever, at least people that expect to have up to date documentation in a accessible way without having to ask and given a folder with loosely structured information.

> Particularly as a website is yet-another-artifact to maintain

That's why it should be all automated so we can commit code and the rest goes on. and then maybe even others can help with making the "website" more nice, I think @wimjongman has experimented in the past with GH pages generation as well.

**Comment by wimjongman** [comment]
Yes, we have the BIRT documentation automated. Just change some markup, PR, merge, and the site is rebuilt.

https://eclipse.org/birt

https://github.com/eclipse/birt-website

**Comment by laeubi** [comment]
@wimjongman that looks really nice! By the way you can:

1. get an own org for bird (actually recommended by eclipse infra)
2. get an nicer domain e.g. we have https://tycho.eclipseprojects.io/ 

just open a ticket in the EF gitlab tracker!

**Comment by wimjongman** [comment]
Thanks, but we have settled on this. It would be disruptive again for the community.

**Comment by laeubi** [comment]
@wimjongman everything is redirected, so there should be no "disruption" but of course it is your choice :-)

**Comment by laeubi** [comment]
I have now started some work here: https://github.com/eclipse-tycho/tycho/pull/1497

**Comment by LorenzoBettini** [comment]
@laeubi sorry, I had forgotten about that... please let me know if you face any problems, and I'll try to help as far as I can.

**Comment by laeubi** [comment]
It seem to work now to generate the sitedoc, but I can't get it to push to another repository in the same organization:

https://github.com/eclipse-tycho/tycho/actions/runs/3218458142/jobs/5262648916

**Comment by laeubi** [comment]
What I could think of is that we simply do it the other way round and put the workflow into the site repo... then it has to be triggered manually but I think that is acceptable and I don't like to play around with setting up PATs and remote triggers and alike ...

**Comment by LorenzoBettini** [comment]
> It seem to work now to generate the sitedoc, but I can't get it to push to another repository in the same organization:
> 
> https://github.com/eclipse-tycho/tycho/actions/runs/3218458142/jobs/5262648916

To do that (it's supported by GitHub Actions), you need some further steps, quoting from another example of mine, https://www.lorenzobettini.it/2021/03/publishing-an-eclipse-p2-composite-repository-on-github-pages/

- create a [GitHub personal access token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) (selecting repo);
- create a secret in the GitHub repository of the project (where we run the GitHub Actions workflow), in this example it is called ACTIONS_TOKEN, with the value of that token;
- when running the Maven deploy command, we need to override the property github-update-repo by specifying a URL for the GitHub repository with the update site using the HTTPS syntax and the encrypted ACTIONS_TOKEN; in this example, it is https://x-access-token:${{ secrets.ACTIONS_TOKEN }}@github.com/LorenzoBettini/p2composite-github-pages-example-updates;
- we also need to configure in advance the Git user and email, with some values, otherwise, Git will complain when creating the commit.

In the steps above (i.e., in my example), the **git push** was performed directly by the POM, that's why it says "Maven deploy", but in your case, it should be put in the workflow directly if I understand correctly. Basically, instead of `actions/upload-artifact@v3` you should have a shell run step to run git push directly on the other repository, by using an URL of the shape:

https://x-access-token:${{ secrets.ACTIONS_TOKEN }}@github.com/<the other repository organization and name>

Unfortunately, I can't try that myself because I don't have write access.

To try these steps, I'd suggest you have a separate workflow that just tries to clone the other repo, create some file, commit and push to the other repo (that's what I did when I wrote that blog post). In particular, once you have the clone URL correctly (with the above shape), everything else should work smoothly (since the remote repo will be already configured when you push).

Please keep me posted

**Comment by LorenzoBettini** [comment]
And in this particular case, I guess the personal access token should be created by the owner of the organization; the owner of the organization does not have to disclose the value of the PAT: the owner can create directly the secret in this repository, without disclosing the value.

**Comment by LorenzoBettini** [comment]
Oh, and I found the experiments I made for that blog post:

https://github.com/LorenzoBettini/p2composite-github-pages-example/blob/master/.github/workflows/experiments.yml

This does the manual git clone, commit, push steps as I was suggesting.

In that example workflow, the destination repository is

`https://x-access-token:${{ secrets.ACTIONS_TOKEN }}@github.com/LorenzoBettini/p2composite-github-pages-example-updates`

If I understand correctly, in your case should be

`https://x-access-token:${{ secrets.ACTIONS_TOKEN }}@github.com/eclipse-tycho/eclipse-tycho.github.io`

**Comment by laeubi** [comment]
I have requested a PAT now here: https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/2061

**Comment by LorenzoBettini** [comment]
As I said, you should also ask them to create a secret in this repository with the PAT value (unless they're Ok with giving you the value of the PAT directly)

**Comment by laeubi** [comment]
I edited the text now and suggested `TYCHO_SITE_PAT` as a name, I think the name itself do not matter much ...

**Comment by LorenzoBettini** [comment]
Yes, the secret's name is not essential. When everything's set up, I suggest you try the experiment, and if that succeeds, you should be good to go with the actual workflow :)

**Comment by laeubi** [comment]
I hope I can even just use the checkout action as described here: 

https://github.com/actions/checkout#push-a-commit-using-the-built-in-token

but with a PAT?

**Comment by laeubi** [comment]
The PAT was created

-  current workflow: https://github.com/eclipse-tycho/tycho/blob/master/.github/workflows/sitedocs.yml
- resulting run: https://github.com/eclipse-tycho/tycho/actions/runs/3235375077

The job succeeds but git push says everything is up to date...

**Comment by laeubi** [comment]
Hurray https://github.com/eclipse-tycho/eclipse-tycho.github.io/tree/main/doc/master

**Comment by wimjongman** [comment]
Hello back.

![image](https://user-images.githubusercontent.com/180969/195398185-bd3a3ca7-482b-47c9-9ac3-323fcca78f2b.png)

**Comment by wimjongman** [comment]
Sure! Can we use Hugo?

**Comment by laeubi** [comment]
The repository is here:

https://github.com/eclipse-tycho/eclipse-tycho.github.io

I think we can use whatever is sufficient here, the only folder currently used is the "doc" folder where we deploy the site-docs but even this can be adjusted if required.

**Comment by HannesWell** [comment]
> Sure! Can we use Hugo?

For curiosity, do you know a (good) Maven plug-in to build with Hugo so that one does not have to install hugo on the build machine?
At work we wanted to build a hugo site that also included asciidoc files. But bcause Since I didn't want to manage hugo and ascidoctor plus its ruby dependencies on our build-server, I ended up writing a Java script (not javascript) that downloads+extracts the hugo zip similar like Jenkins automated tool installer as well as the `ascidoctorJ` zip and then creates a little shell/bat script that tricks the hugo binary to use `asciidoctorJ` instead of `asciidoctor`.
I know asciidoctor is not (yet) planned to be used by Tycho, nevertheless I'm curious if you know a more 'Maven' native method?
In case such Java script is interesting for you/Tycho, I can contribute that where needed.

**Comment by laeubi** [comment]
At least for Github there seems to be an action to do so:
https://gohugo.io/hosting-and-deployment/hosting-on-github/#build-hugo-with-github-action

**Comment by LorenzoBettini** [comment]
Just for curiosity, why another framework? Isn't markdown or other format supported by the Maven site plugin enough?

**Comment by wimjongman** [comment]
> Just for curiosity, why another framework? Isn't markdown or other format supported by the Maven site plugin enough?

Is that not another framework as well?

**Comment by wimjongman** [comment]
> > Sure! Can we use Hugo?
> 
> For curiosity, do you know a (good) Maven plug-in ...

I don't know that, but I know that the foundation supports Hugo. 

* https://www.google.com/search?q=eclipse+foundation+hugo

**Comment by LorenzoBettini** [comment]
> Is that not another framework as well?

Yes, but it's already integrated with Maven and, most of all, if I understand correctly, the site will contain the Tycho Maven goals documentation, which is already dealt with by Maven.

**Comment by akurtakov** [comment]
I have no idea what hugo is (even from that google search) and as Tycho is a maven plugin it's probably best to just deploy maven side.

**Comment by LorenzoBettini** [comment]
That's exactly what I meant. During the Maven build the HTML is produced and can just be published as it is.

**Comment by HannesWell** [comment]
> I have no idea what hugo is (even from that google search) and as Tycho is a maven plugin it's probably best to just deploy maven side.

It is a static website generator Framework:
https://gohugo.io/

Personally I mixed feelings about, it works well when you do the standard stuff (especially when building it) and has wide theming support and the sites look nice. But I had a hard time to integrate it into a Maven build.

Also integrating the output of the maven-site plugin could become difficult. But maybe we are lucky.

**Comment by wimjongman** [comment]
> > Is that not another framework as well?
> 
> Yes, but it's already integrated with Maven and, most of all, if I understand correctly, the site will contain the Tycho Maven goals documentation, which is already dealt with by Maven.

No that is already been taking care of here: https://tycho.eclipseprojects.io/doc/3.0.0/tycho-packaging-plugin/plugin-info.html

It will be the official Tycho landing page. I admit it is hard to beat the current site: https://tycho.eclipseprojects.io/ :)

> Personally I mixed feelings about, it works well when you do the standard stuff (especially when building it)

It is supported by the EF

**Comment by laeubi** [comment]
As long as one can edit a file in the repository and it automatically updates the website afterwards automatically I don't think there is any need for a maven-build as this is completely independent from the code (we already have maven-site-build for the "technical" part, but we could use a little nice landing pages, and there I won't mind the technique used. SO if aynone wants to contribute something the technique does not matter much as long as it is not handcrafted HTML :-)

I think Gihub even supports using plain Markdown files and it seems you can [theme it using jekyll](https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/adding-a-theme-to-your-github-pages-site-using-jekyll), but in the end it doesn't hlep if we have the "perfect framework" but no one creates some content in that framework :-)

> Also integrating the output of the maven-site plugin could become difficult. But maybe we are lucky.

A simple link would be sufficient if nothing else is supported...
